File: MtCoffee.Web/ClientApp/src/models/product.ts

export type DrinkSizeOption = '8oz' | '12oz' | '16oz' | '20oz' | '24oz' | '32oz';
// export type ProductCategory = 'Drinks' | 'Food';
// export type ProductSubcategory = 'Hot/Iced' | 'Bubble Tea' | 'Energy' | 'Blended' | 'Tea and Other' | 'Pastry'; 

export interface Product {
    id?: number;
    name: string;
    description?: string;
    price: number;
    isActive: boolean;

    categoryId?: number;
    hasVariants?: Readonly<boolean>;  // readonly
    discountIds: number[];
    productOptionIds: number[];
    variants: ProductVariant[];
}

export interface ProductVariant {
    id?: number;
    productId: number;
    name: string;
    priceOverride?: number;
    isDefaultVariant: boolean;
}

export interface ProductOption {
    id?: number;
    name: string;
    price?: number;
    optionGroupKey?: string;
    isQuantityEnabled: boolean;
    isActive: boolean;
}

export interface ProductCategory {
    id?: number;
    name: string;
    tileColor: string;
    sortOrder: number;
    categoryGroupName: string;
}

File: MtCoffee.Web/Models/Product/Product.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace MtCoffee.Web.Models.Product
{
    public class Product
    {
        public int? Id { get; set; }
        [Required]
        public string Name { get; set; }
        public string Description { get; set; }
        public bool IsActive { get; set; }
        public double Price { get; set; }

        public int? CategoryId { get; set; }

        public virtual bool HasVariants { get => Variants.Any(); }

        public virtual ICollection<int> DiscountIds { get; set; }
        public virtual ICollection<ProductVariant> Variants { get; set; } = new List<ProductVariant>();
        public virtual ICollection<int> ProductOptionIds { get; internal set; }
    }

    public class ProductVariant
    {
        public int? Id { get; set; }
        [Required]
        public string Name { get; set; }
        public int ProductId { get; set; }
        public double? PriceOverride { get; set; } = null;
        public bool IsDefaultVariant { get; set; } = false;
    }
}
